aztk_cli/spark/endpoints/cluster/cluster_ssh.py

```diff
@@ -88,16 +73,72 @@ def print_plugin_ports(cluster_config: ClusterConfiguration):
                 if port.expose_publicly:
                     has_ports = True
                     plugin_ports[plugin.name].append(port)
-        
+
         if has_ports:
             log.info("plugins:")
 
         for plugin in plugin_ports:
             if plugin_ports[plugin]:
-                log.info("  " + plugin)
+                log.info(" %s ", plugin)
                 for port in plugin_ports[plugin]:
                     label = "    - open"
                     if port.name:
                         label += " {}".format(port.name)
                     url = "{0}{1}".format(http_prefix, port.public_port)
                     utils.log_property(label, url)
+
+
+def native_python_ssh_into_master(spark_client, cluster, ssh_conf, password):
+    if not ssh_conf.connect:
+        log.warning("No ssh client found, using pure python connection.")
+        return
+
+    configuration = spark_client.get_cluster_config(cluster.id)
+    plugin_ports = []
+    if configuration and configuration.plugins:
+        ports = [
+            PortForwardingSpecification(
+                port.internal,
+                port.public_port) for plugin in configuration.plugins for port in plugin.ports if port.expose_publicly
+        ]
+        plugin_ports.extend(ports)
+
+    print("Press ctrl+c to exit...")
+    spark_client.cluster_ssh_into_master(
+        cluster.id,
+        cluster.master_node_id,
+        ssh_conf.username,
+        ssh_key=None,
+        password=password,
+        port_forward_list=[
+            PortForwardingSpecification(remote_port=8080, local_port=8080),      # web ui
+            PortForwardingSpecification(remote_port=4040, local_port=4040),      # job ui
+            PortForwardingSpecification(remote_port=18080, local_port=18080),    # job history ui
+        ] + plugin_ports,
+        internal=ssh_conf.internal
+    )
+
+
+def shell_out_ssh(spark_client, ssh_conf):
+    try:
+        ssh_cmd = utils.ssh_in_master(
+            client=spark_client,
+            cluster_id=ssh_conf.cluster_id,
+            webui=ssh_conf.web_ui_port,
+            jobui=ssh_conf.job_ui_port,
+            jobhistoryui=ssh_conf.job_history_ui_port,
+            username=ssh_conf.username,
+            host=ssh_conf.host,
+            connect=ssh_conf.connect,
+            internal=ssh_conf.internal)
+
+        if not ssh_conf.connect:
+            log.info("")
+            log.info("Use the following command to connect to your spark head node:")
+            log.info("\t%s", ssh_cmd)
+
+    except batch_error.BatchErrorException as e:
+        if e.error.code == "PoolNotFound":
```

Comment:
I think we should have an enum(or const) for those error codes

Reply:
agreed, but I think that should probably come in a single PR for the entire repo.